### DISCUSSION TITLE
Allow a way to use a custom token verification flow

### DISCUSSION THREAD

**Discussion by viters** [directus/directus#8727] [Feature Requests]
I've recently came into need of implementing extension of Directus authentication mechanism, where I would accept JWT tokens coming from my Identity provider (signed with JWK). I developed a solution hook, that hooks into `middlewares.init.before`, but I had to workaround setting `req.accountability` with:

```js
Object.defineProperty(req, 'accountability', {
  get() {
    return {
      user: user.id,
      role: user.role,
      admin: false,
      app: false,
      ip: req.ip.startsWith('::ffff:') ? req.ip.substring(7) : req.ip,
      userAgent: req.get('user-agent'),
    };
  },
  set(newValue) {},
});
```

because otherwise `authenticate` middleware would wipe out my session information.

**Comment by rijkvanzanten** [comment]
The authenticate middleware is currently called for every request:

https://github.com/directus/directus/blob/b2ec1824934bde7396fe76b4154f1db2c99344c4/api/src/app.ts#L154

and will construct the accountability object fresh based on the passed JWT or static token:

https://github.com/directus/directus/blob/b2ec1824934bde7396fe76b4154f1db2c99344c4/api/src/middleware/authenticate.ts#L12-L82

In order to allow for this particular use case, we need to have a way to mix in a custom authentication check handler.

We could consider adding a new "authenticate" filter hook that allows for implementing any custom authentication check. With that in place, we could do something like

```js
const accountability = await emitFilter('authenticate');

if (accountability) {
  req.accountability = accountability;
  return next();
} else {
  // do what we have now
}
```

Thoughts @nickrum?

**Reply by nickrum** [reply]
Sound like a useful thing to add. It would be nice if we could pass the `accountability` object constructed by the API to the filter and then set `req.accountability` to whatever is returned by the filter, but it doesn't look like this is easily possible with the current structure of the authenticate middleware.

**Reply by rijkvanzanten** [reply]
That's something we can fairly easily work in I think 👍🏻 

It's mostly the problem around the token override. We're effectively saying "Skip any token check in favor of our custom middleware"

**Comment by viters** [comment]
I agree completely with ignoring whole `authenticate` middleware, when `req.accountability` is defined.

You are right in https://github.com/directus/directus/pull/8691#issuecomment-941114295
> This approach will still check the user password / token right afterwards, and will throw a InvalidCredentialsException when the user isn't authenticating through Directus' means 🤔

My solution at #8691 was only aimed at unavoidable redefinition of `req.accountability`, because it is easier to fix without any logic changes. Although it's true, that sole existence of `token` will lead to `InvalidCredentialsException`. That's why in my middleware, I had to clear tokens from request, before it reaches `authenticate` middleware:

```js
Object.defineProperty(req, 'accountability', {
  get() {
    return {
      user: user.id,
      role: user.role,
      admin: false,
      app: false,
      ip: req.ip.startsWith('::ffff:') ? req.ip.substring(7) : req.ip,
      userAgent: req.get('user-agent'),
    };
  },
  set(newValue) {},
});

if (req.query && req.query.access_token) {
  req.query.access_token = undefined;
}

if (req.headers && req.headers.authorization) {
  req.headers.authorization = '';
}
```

If `authenticate` middleware would start with
```ts
if (req.accountability != null) return next();
```
it solves both problems.

**Reply by rijkvanzanten** [reply]
Don't think we can just add that as a check. It feels risky, as that req object can be modified from any other hook/middleware. I'd rather err on the side of explicitness, and rely on a hook-based override that makes it an opt-in.

**Reply by viters** [reply]
Risky, because? What kind of problem do you see that may arise? Is it risky internally (regarding future api module development) or externally (because people may bypass `authenticate`)? If you mean external risk - well that's exactly what I want to do - bypass `authenticate` in specific situations - and I see no more risk to it than modifying `app` itself.

Just to be clear, opt-in/out should be possible based on request. In other words, right now I allow to call Directus API both to people with Directus tokens, and JWK-signed tokens coming from my other service. It should not be "enable/disable for good".

**Reply by rijkvanzanten** [reply]
It feels like it might be a foot-gun. Even something as simple as setting req.accountability = false would immediately disable all auth in this middleware based on a check like that.

> Just to be clear, opt-in/out should be possible based on request. In other words, right now I allow to call Directus API both to people with Directus tokens, and JWK-signed tokens coming from my other service. It should not be "enable/disable for good".

I'm aware. My suggestion is just to rely on a dedicated hook, instead of the default value of a property on req 👍🏻

**Reply by viters** [reply]
> Even something as simple as setting req.accountability = false would immediately disable all auth in this middleware based on a check like that.

Yeah, that's true. I find it completely reasonable - `accountability` is Directus session - if you override it, you opt out from Directus session creation. There should also be documented what is the minimal `accountability`, required by other parts of Directus.

> I'm aware. My suggestion is just to rely on a dedicated hook, instead of the default value of a property on req 👍🏻

Okay, gotcha. To be honest - does it really make a difference, whether I set `accountability` in `middlewares.before` hook or in `authenticate.before` hook? In both cases you can set `accountability = false`. If you are afraid of stuff that people can set in `accountability`, it would be better to type it and / or verify if user has set it correctly.

```ts
if (req.accountability != null) {
  // validate accountability, throw if missing required properties
}
```

**Reply by rijkvanzanten** [reply]
> To be honest - does it really make a difference, whether I set accountability in middlewares.before hook or in authenticate.before hook?

Yes, you won't touch req.accountability directly in the latter case.

**Reply by viters** [reply]
Okay, you have a point. I see it, I agree.

So, something like this?

```ts
req.accountability = {
	user: null,
	role: null,
	admin: false,
	app: false,
	ip: req.ip.startsWith('::ffff:') ? req.ip.substring(7) : req.ip,
	userAgent: req.get('user-agent'),
};

const accountability = await emitAsyncSafe('authentication', {
	accountability: req.accountability,
	req,
	token: req.token,
});

if (accountability != null) {
	req.accountability = accountability;
	return next();
}

if (!req.token) return next();

// ...default authentication
```

**Reply by rijkvanzanten** [reply]
Yeah something along those lines 👍🏻

**Reply by viters** [reply]
I've created a PR with this, #8750

**Comment by sjones6** [comment]
Note: released as part of v9.6.0: https://github.com/directus/directus/releases/tag/v9.6.0

PR https://github.com/directus/directus/pull/11942
